runtime/kusama/src/lib.rs

```diff
@@ -2130,6 +2132,43 @@ sp_api::impl_runtime_apis! {
 						fun: Fungible(1_000_000 * UNITS),
 					}].into()
 				}
+				fn ensure_for_send(origin_ref: &MultiLocation, _dest: &MultiLocation, fee_reason: FeeReason) -> Option<FeesMode> {
+					use xcm_executor::traits::{FeeManager, TransactAsset};
+
+					let mut fees_mode = None;
+					if !<XcmConfig as xcm_executor::Config>::FeeManager::is_waived(Some(origin_ref), fee_reason) {
+						// if not waived, we need to set up accounts for paying and receiving fees
+
+						// mint ED to origin
+						let ed = MultiAsset {
+							id: Concrete(TokenLocation::get()),
+							fun: Fungible(EXISTENTIAL_DEPOSIT.into())
+						};
+						<XcmConfig as xcm_executor::Config>::AssetTransactor::deposit_asset(&ed, &origin_ref, None).unwrap();
+
+						// overestimate delivery fee
+						use runtime_common::xcm_sender::PriceForParachainDelivery;
+						let overestimated_xcm = vec![ClearOrigin; 128].into();
+						let overestimated_fees = runtime_common::xcm_sender::ExponentialPrice::<
+							xcm_config::FeeAssetId,
+							xcm_config::BaseDeliveryFee,
+							TransactionByteFee,
+							Dmp
+						>::price_for_parachain_delivery(
+							kusama_runtime_constants::system_parachain::ASSET_HUB_ID.into(),
+							&overestimated_xcm
+						);
+
+						// mint overestimated fee to origin
+						for fee in overestimated_fees.inner() {
+							<XcmConfig as xcm_executor::Config>::AssetTransactor::deposit_asset(&fee, &origin_ref, None).unwrap();
+						}
+
+						// expected worst case
+						fees_mode = Some(FeesMode { jit_withdraw: true });
+					}
+					fees_mode
+				}
```

Comment:
This (rather complex) code will likely need to be copy-pasted across every single Polkadot/Kusama chain. Therefore, best to keep it in some common place and have a simple means of calling into it with some appropriate parameterisation.

Reply:
ok, thank you, extracted to reusable code, which will be also reused for system parachains